### DISCUSSION TITLE
Webpack does not respect pnpm workspace peerDependencies

### DISCUSSION THREAD

**Discussion by kbrooks** [webpack/webpack#18192] [General]
I ran into an issue where webpack does not resolve peerDependcies correctly on a pnpm monorepo, leading to duplicate packages being loaded. 

Is this a bug in webpack?

I reproduced the issue here: https://github.com/kbrooks/peer-dep-problem-repro

The repo packs two versions of jotai into `test-app` instead of using the peerDependencies  field on `test-package`'s package.json to use the same version of jotai used by `test-app`.

**Comment by alexander-akait** [comment]
I think you are looking for https://github.com/webpack/enhanced-resolve/issues/410

**Reply by alexander-akait** [reply]
You can find a workaround there, try to check it out

**Comment by kbrooks** [comment]
By default, pnpm workspaces do not work with peer dependencies because of how pnpm uses symlinks to create workspace dependencies:
ls -l packages/my-app/node_modules/@shared

build -> ../../../build
jest -> ../../../jest
...etc
So instead of symlinking to [root]/node_modules/.pnpm/[dependency] like a package installed from the registry, pnpm just defines a symlink directly into the workspace. 
Because dev dependencies are installed by pnpm i to every workspace, they are present in the symlinked node_modules, and will be resolved by webpack and jest in imports from that workspace.

The [Pnpm dependenciesMeta.injected rule](https://pnpm.io/package_json#dependenciesmetainjected) makes workspace dependencies install like they came from the registry. All you need to do is set injected to true under dependenciesMeta in the package.json that declares the workspace dependency:

"dependenciesMeta": {
      "[package-name]": {
      "injected": true
}

This tells pnpm to install the workspace dependency under node-modules/.pnpm/file+[package-specifier] instead of symlinking to the workspace directory. 
ls -l test-app/node_modules
test-package -> ../../../node_modules/.pnpm/file+packages+test-package_jotai@2.7.0_react@18.2.0/node_modules/test-package

Because the dependency is installed as a normal node package, it does not include dev dependencies and respects the peer dependency configuration. However, because you lose the symlink, you lose the ability to make local changes and have them be reflected immediately. The workspace dependency under node_modules/.pnpm is a copy of the files in packages/[package-name], so it will not update until you run pnpm i.
